align/pnr/toplevel.py

```diff
@@ -626,7 +626,10 @@ def place_and_route(*, DB, opath, fpath, numLayout, effort, adr_mode, PDN_mode,
 
                 for constraint in frontier[top_name]:
                     assert constraint['constraint'] == 'place_on_grid'
-                    assert constraint['ored_terms'], f'No legal grid locations for {top_name} {constraint}'
```

Comment:
@stevenmburns , this assertion is downgraded to a warning so that a human-readable error message is printed in `checker.py` at the end of the flow.

Reply:
I guess that's better. Or we could report something readable here.

Reply:
The problem might be that can't continue to generate hierarchical placements because the placement constraints are unsatisfiable. What should we do in this case? We could remove that placement and hope that we get another variant that is usable. But then the placer shouldn't have generated it in the first place. I thought is should be an assertion here because it shouldn't happen. But it could be an overconstrained input error. But then we should report that, that than a placement was returned that is wrong. That is why hardening the constraint and checking it before the placer is run is as important to do here as it is for other constraints. Maybe the checking before placement is just as hard as doing the placement. If that is the case we might have to reconsider this way.

Reply:
I agree that checking all possible combinations of placement against constraints is not a trivial problem.
Examples for horizontal grid constraints:
`{"constraint": "Align", "line": "h_bottom", "instances": ["i0", "i1"]}` 
=> There must exist at least one pair of concrete versions of i0 and i1 that can be aligned horizontally.

`{"constraint": "Order", "direction": "top_to_bottom", "instances": ["i1", "i2"], "abut": True}` 
=> There must exist at least one pair of concrete versions of i1 and i2 that can be abutted.

`{"constraint": "SymmetricBlocks", "direction": "V", "pairs": [["i2", "i3"]]`
=> There must exist at least one pair of concrete versions of i2 and i3 that have the same center of gravity when placed.

Theoretically, above could still be checked for leaf blocks by introducing decision variable per concrete name and adding formulas for legal bbox values.

Yet handling the the constraints on intermediate hierarchy is more complicated as all possible concretization of the subhierarchy need to be enumerated.